File: data/scripts/editor.py
import json
import os
from flask import Flask, render_template_string, request, jsonify, redirect, url_for, send_from_directory

SCRIPT_DIR = os.path.dirname(os.path.abspath(__file__))
app = Flask(__name__, static_folder=SCRIPT_DIR, static_url_path="/static")

DATA_DIR = os.path.normpath(os.path.join(os.path.dirname(__file__), ".."))

SCHEMA = {
    "films": {
        "pk": "film_id",
        "fields": {
            "film_id": {"type": "text", "required": True},
            "film_name": {"type": "text", "required": True},
            "film_iso": {"type": "number", "required": True},
            "film_description": {"type": "text", "required": False},
            "film_type": {"type": "select", "required": True, "options": [
                "black & white", "color (C-41)", "color (E-6)", "special"
            ]},
            "film_popularity": {"type": "number", "required": False},
            "film_datasheet": {"type": "text", "required": False},
            "film_lomography_id": {"type": "text", "required": False},
            "film_flickr_search": {"type": "text", "required": False},
        },
    },
    "sizes": {
        "pk": "size_id",
        "fields": {
            "size_id": {"type": "text", "required": True, "computed": ["film_id", "size_format"]},
            "film_id": {"type": "fk", "required": True, "ref": "films"},
            "size_format": {"type": "select", "required": True, "options": [
                "110", "120", "127", "135", "4x5", "8x10"
            ]},
            "size_year": {"type": "number", "required": False},
        },
    },
    "packs": {
        "pk": "pack_id",
        "fields": {
            "pack_id": {"type": "text", "required": True, "computed": ["size_id", "pack_count"]},
            "size_id": {"type": "fk", "required": True, "ref": "sizes"},
            "pack_count": {"type": "number", "required": True},
        },
    },
    "stores": {
        "pk": "store_id",
        "fields": {
            "store_id": {"type": "text", "required": True},
            "store_name": {"type": "text", "required": True},
            "store_url": {"type": "text", "required": True},
            "store_culture": {"type": "text", "required": False},
            "store_price_selector": {"type": "text", "required": False},
            "store_availability_selector": {"type": "text", "required": False},
        },
    },
    "pages": {
        "pk": None,
        "fields": {
            "pack_id": {"type": "fk", "required": True, "ref": "packs"},
            "store_id": {"type": "fk", "required": True, "ref": "stores"},
            "page_url": {"type": "text", "required": True},
        },
    },
}


def load_data(table):
    path = os.path.join(DATA_DIR, f"{table}.json")
    with open(path, "r", encoding="utf-8") as f:
        return json.load(f)


def save_data(table, data):
    path = os.path.join(DATA_DIR, f"{table}.json")
    with open(path, "w", encoding="utf-8") as f:
        json.dump(data, f, indent=2, ensure_ascii=False)
        f.write("\n")


def get_display_name(table, row):
    schema = SCHEMA[table]
    pk = schema["pk"]
    if pk:
        return row[pk]
    fields = list(schema["fields"].keys())
    return " | ".join(str(row.get(f, "")) for f in fields[:2])


# --- API ---

@app.route("/api/<table>", methods=["GET"])
def api_list(table):
    if table not in SCHEMA:
        return jsonify({"error": "Unknown table"}), 404
    return jsonify(load_data(table))


@app.route("/api/<table>", methods=["POST"])
def api_create(table):
    if table not in SCHEMA:
        return jsonify({"error": "Unknown table"}), 404
    data = load_data(table)
    row = request.json
    schema = SCHEMA[table]
    for field, spec in schema["fields"].items():
        if spec["type"] == "number" and field in row and row[field] is not None:
            try:
                row[field] = int(row[field])
            except (ValueError, TypeError):
                pass
        if row.get(field) == "" and not spec["required"]:
            row[field] = None
    data.append(row)
    pk = schema["pk"]
    if pk:
        data.sort(key=lambda r: r.get(pk, ""))
    save_data(table, data)
    return jsonify({"ok": True})


@app.route("/api/<table>/<int:index>", methods=["PUT"])
def api_update(table, index):
    if table not in SCHEMA:
        return jsonify({"error": "Unknown table"}), 404
    data = load_data(table)
    if index < 0 or index >= len(data):
        return jsonify({"error": "Index out of range"}), 404
    row = request.json
    schema = SCHEMA[table]
    for field, spec in schema["fields"].items():
        if spec["type"] == "number" and field in row and row[field] is not None:
            try:
                row[field] = int(row[field])
            except (ValueError, TypeError):
                pass
        if row.get(field) == "" and not spec["required"]:
            row[field] = None
    data[index] = row
    save_data(table, data)
    return jsonify({"ok": True})


@app.route("/api/<table>/<int:index>", methods=["DELETE"])
def api_delete(table, index):
    if table not in SCHEMA:
        return jsonify({"error": "Unknown table"}), 404
    data = load_data(table)
    if index < 0 or index >= len(data):
        return jsonify({"error": "Index out of range"}), 404
    data.pop(index)
    save_data(table, data)
    return jsonify({"ok": True})


@app.route("/api/fk_options/<table>")
def api_fk_options(table):
    if table not in SCHEMA:
        return jsonify({"error": "Unknown table"}), 404
    data = load_data(table)
    pk = SCHEMA[table]["pk"]
    if not pk:
        return jsonify([])
    return jsonify(sorted(set(row[pk] for row in data)))



# --- UI ---

TEMPLATE = r"""
<!DOCTYPE html>
<html lang="en">
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1">
<title>analog films</title>
<link rel="stylesheet" href="/static/oat.min.css">
<link rel="stylesheet" href="/static/editor.css">
<script src="/static/oat.min.js" defer></script>
</head>
<body data-sidebar-layout data-table="{{ table }}">

<aside data-sidebar>
  <header><strong>analog films</strong></header>
  <nav>
    <ul>
      {% for t in tables %}
      <li><a href="/{{ t }}" {{ 'aria-current="page"' | safe if t == table else '' }}>{{ t }}</a></li>
      {% endfor %}
    </ul>
  </nav>
  <footer>
    <label>
      <input type="checkbox" role="switch" id="theme-toggle" onchange="toggleTheme()">
      Dark mode
    </label>
  </footer>
</aside>

<main style="padding: var(--space-3);">
  <div class="toolbar">
    <h2>{{ table }} <span class="count" id="count"></span></h2>
    <fieldset class="group">
      <input type="search" id="search" placeholder="Filter rows...">
      <button onclick="openCreate()">+ Add</button>
    </fieldset>
  </div>

  <div class="table-wrap">
    <table>
      <thead><tr id="thead"></tr></thead>
      <tbody id="tbody"></tbody>
    </table>
  </div>
</main>

<dialog id="modal" closedby="any">
  <form method="dialog" id="form" onsubmit="return handleSubmit(event)">
    <header><h3 id="modal-title">Add</h3></header>
    <div id="form-fields"></div>
    <footer>
      <button type="button" class="outline" onclick="closeModal()">Cancel</button>
      <button type="submit" id="submit-btn">Save</button>
    </footer>
  </form>
</dialog>

<script id="schema-data" type="application/json">{{ schema_json | safe }}</script>
<script src="/static/editor.js"></script>
</body>
</html>
"""


@app.route("/")
def index():
    return redirect(url_for("table_view", table="films"))


@app.route("/<table>")
def table_view(table):
    if table not in SCHEMA:
        return "Not found", 404
    return render_template_string(
        TEMPLATE,
        table=table,
        tables=list(SCHEMA.keys()),
        schema_json=json.dumps(SCHEMA),
    )


if __name__ == "__main__":
    print(f"Data directory: {DATA_DIR}")
    print(f"Opening editor at http://127.0.0.1:5000")
    app.run(host="127.0.0.1", port=5000, debug=True)

File: data/scripts/editor.js
const TABLE = document.body.dataset.table;
const SCHEMA = JSON.parse(document.getElementById("schema-data").textContent);
const tableSchema = SCHEMA[TABLE];
const fields = Object.keys(tableSchema.fields);
let rows = [];
let editIndex = null;
const fkCache = {};

async function loadFkOptions(refTable) {
  if (fkCache[refTable]) return fkCache[refTable];
  const r = await fetch(`/api/fk_options/${refTable}`);
  const opts = await r.json();
  fkCache[refTable] = opts;
  return opts;
}

async function loadData() {
  const r = await fetch(`/api/${TABLE}`);
  rows = await r.json();
  render();
}

function render() {
  const q = document.getElementById("search").value.toLowerCase();
  const filtered = q ? rows.filter(r => JSON.stringify(r).toLowerCase().includes(q)) : rows;
  document.getElementById("count").textContent = `(${filtered.length}${q ? ` of ${rows.length}` : ''})`;

  let th = fields.map(f => `<th>${f}</th>`).join("") + "<th>Actions</th>";
  document.getElementById("thead").innerHTML = th;

  if (filtered.length === 0) {
    document.getElementById("tbody").innerHTML = `<tr><td colspan="${fields.length+1}">No rows found</td></tr>`;
    return;
  }
  let html = "";
  for (const row of filtered) {
    const realIdx = rows.indexOf(row);
    let cells = fields.map(f => {
      let v = row[f];
      if (v === null || v === undefined) return `<td><span class="badge">null</span></td>`;
      return `<td title="${String(v).replace(/"/g, '&quot;')}">${escHtml(String(v))}</td>`;
    }).join("");
    cells += `<td class="actions">
      <button class="small outline" onclick="openEdit(${realIdx})">Edit</button>
      <button class="small" data-variant="danger" onclick="confirmDelete(${realIdx})">Delete</button>
    </td>`;
    html += `<tr>${cells}</tr>`;
  }
  document.getElementById("tbody").innerHTML = html;
}

function escHtml(s) {
  return s.replace(/&/g,"&amp;").replace(/</g,"&lt;").replace(/>/g,"&gt;").replace(/"/g,"&quot;");
}

function fuzzyMatch(text, query) {
  const t = text.toLowerCase();
  const terms = query.toLowerCase().split(/\s+/).filter(Boolean);
  return terms.every(term => t.includes(term));
}

function highlightMatch(text, query) {
  const escaped = escHtml(text);
  const terms = query.toLowerCase().split(/\s+/).filter(Boolean);
  if (!terms.length) return escaped;
  const lower = escaped.toLowerCase();
  const ranges = [];
  for (const term of terms) {
    let start = 0;
    while (true) {
      const idx = lower.indexOf(term, start);
      if (idx === -1) break;
      ranges.push([idx, idx + term.length]);
      start = idx + 1;
    }
  }
  if (!ranges.length) return escaped;
  ranges.sort((a, b) => a[0] - b[0]);
  const merged = [ranges[0]];
  for (let i = 1; i < ranges.length; i++) {
    const last = merged[merged.length - 1];
    if (ranges[i][0] <= last[1]) {
      last[1] = Math.max(last[1], ranges[i][1]);
    } else {
      merged.push(ranges[i]);
    }
  }
  let result = '', pos = 0;
  for (const [s, e] of merged) {
    result += escaped.slice(pos, s) + '<mark>' + escaped.slice(s, e) + '</mark>';
    pos = e;
  }
  return result + escaped.slice(pos);
}

function hideFkDropdown(id) {
  document.getElementById(id + '-list').style.display = 'none';
}

function showFkDropdown(id) {
  filterFkDropdown(id);
  document.getElementById(id + '-list').style.display = '';
}

function filterFkDropdown(id) {
  const input = document.getElementById(id);
  const list = document.getElementById(id + '-list');
  const q = input.value;
  let visibleCount = 0;
  for (const li of list.children) {
    li.classList.remove('active');
    const val = li.getAttribute('data-value');
    if (!q || fuzzyMatch(val, q)) {
      li.classList.remove('hidden');
      li.innerHTML = q ? highlightMatch(val, q) : escHtml(val);
      visibleCount++;
    } else {
      li.classList.add('hidden');
    }
  }
  list.style.display = visibleCount ? '' : 'none';
}

function getVisibleItems(id) {
  const list = document.getElementById(id + '-list');
  return Array.from(list.children).filter(li => !li.classList.contains('hidden'));
}

function fkKeydown(e, id) {
  const list = document.getElementById(id + '-list');
  if (list.style.display === 'none') return;
  const items = getVisibleItems(id);
  if (!items.length) return;
  const current = items.findIndex(li => li.classList.contains('active'));
  if (e.key === 'ArrowDown') {
    e.preventDefault();
    const next = current < items.length - 1 ? current + 1 : 0;
    items.forEach(li => li.classList.remove('active'));
    items[next].classList.add('active');
    items[next].scrollIntoView({ block: 'nearest' });
  } else if (e.key === 'ArrowUp') {
    e.preventDefault();
    const next = current > 0 ? current - 1 : items.length - 1;
    items.forEach(li => li.classList.remove('active'));
    items[next].classList.add('active');
    items[next].scrollIntoView({ block: 'nearest' });
  } else if (e.key === 'Enter') {
    if (current >= 0) {
      e.preventDefault();
      pickFk(id, items[current].getAttribute('data-value'));
    }
  }
}

function pickFk(id, value) {
  const input = document.getElementById(id);
  input.value = value;
  input.dispatchEvent(new Event('change'));
  document.getElementById(id + '-list').style.display = 'none';
}

document.addEventListener('mousedown', function(e) {
  const li = e.target.closest('.fk-dropdown li[data-fk]');
  if (li) {
    e.preventDefault();
    pickFk(li.getAttribute('data-fk'), li.getAttribute('data-value'));
    return;
  }
  document.querySelectorAll('.fk-dropdown').forEach(ul => {
    if (!ul.parentElement.contains(e.target)) ul.style.display = 'none';
  });
});

document.getElementById('modal').addEventListener('scroll', function() {
  document.querySelectorAll('.fk-dropdown').forEach(ul => { ul.style.display = 'none'; });
});

document.getElementById('modal').addEventListener('keydown', function(e) {
  if (e.key === 'Escape') {
    const open = document.querySelectorAll('.fk-dropdown:not([style*="display: none"])');
    if (open.length) {
      e.stopPropagation();
      e.preventDefault();
      open.forEach(ul => { ul.style.display = 'none'; });
    }
  }
});

async function buildForm(row) {
  let html = "";
  for (const [field, spec] of Object.entries(tableSchema.fields)) {
    const val = row ? (row[field] ?? "") : "";
    const req = spec.required ? "required" : "";
    if (spec.type === "select") {
      html += `<div data-field><label>${field}${spec.required ? ' *' : ''}</label>`;
      html += `<select name="${field}" ${req}>`;
      if (!spec.required) html += `<option value="">--</option>`;
      for (const o of spec.options) {
        html += `<option value="${o}" ${val === o ? "selected" : ""}>${o}</option>`;
      }
      html += `</select></div>`;
    } else if (spec.type === "fk") {
      const opts = await loadFkOptions(spec.ref);
      const fkId = `fk-${field}`;
      html += `<div data-field><label>${field}${spec.required ? ' *' : ''}</label>`;
      html += `<div style="position:relative;">`;
      html += `<input type="text" name="${field}" id="${fkId}" value="${escHtml(String(val))}" ${req} autocomplete="off"
        onfocus="showFkDropdown('${fkId}')" oninput="filterFkDropdown('${fkId}')" onkeydown="fkKeydown(event, '${fkId}')" onblur="hideFkDropdown('${fkId}')">`;
      html += `<ul id="${fkId}-list" class="fk-dropdown" style="display:none;">`;
      for (const o of opts) {
        html += `<li data-value="${escHtml(o)}" data-fk="${fkId}">${escHtml(o)}</li>`;
      }
      html += `</ul></div></div>`;
    } else if (spec.type === "number") {
      html += `<label data-field>${field}${spec.required ? ' *' : ''}
        <input type="number" name="${field}" value="${val}" ${req}>
      </label>`;
    } else if (field.includes("description")) {
      html += `<label data-field>${field}${spec.required ? ' *' : ''}
        <textarea name="${field}" ${req}>${escHtml(String(val))}</textarea>
      </label>`;
    } else {
      const ro = spec.computed ? 'readonly disabled' : '';
      html += `<label data-field>${field}${spec.required ? ' *' : ''}
        <input type="text" name="${field}" value="${escHtml(String(val))}" ${req} ${ro}>
      </label>`;
    }
  }
  document.getElementById("form-fields").innerHTML = html;
  bindComputedFields();
}

function bindComputedFields() {
  for (const [field, spec] of Object.entries(tableSchema.fields)) {
    if (!spec.computed) continue;
    const target = document.querySelector(`[name="${field}"]`);
    if (!target) continue;
    const update = () => {
      const parts = spec.computed.map(src => {
        const el = document.querySelector(`[name="${src}"]`);
        return el ? el.value : '';
      });
      if (parts.every(p => p)) {
        target.value = parts.join('-');
      } else {
        target.value = '';
      }
    };
    for (const src of spec.computed) {
      const el = document.querySelector(`[name="${src}"]`);
      if (el) {
        el.addEventListener('input', update);
        el.addEventListener('change', update);
      }
    }
    update();
  }
}

function openCreate() {
  editIndex = null;
  document.getElementById("modal-title").textContent = "Add " + TABLE;
  document.getElementById("submit-btn").textContent = "Create";
  buildForm(null);
  document.getElementById("modal").showModal();
}

function openEdit(idx) {
  editIndex = idx;
  document.getElementById("modal-title").textContent = "Edit " + TABLE;
  document.getElementById("submit-btn").textContent = "Update";
  buildForm(rows[idx]);
  document.getElementById("modal").showModal();
}

function closeModal() {
  document.getElementById("modal").close();
}

async function handleSubmit(e) {
  e.preventDefault();
  document.querySelectorAll('#form [disabled]').forEach(el => { el.disabled = false; });
  const fd = new FormData(document.getElementById("form"));
  const row = {};
  for (const [field, spec] of Object.entries(tableSchema.fields)) {
    let v = fd.get(field);
    if (spec.type === "number" && v !== "" && v !== null) v = Number(v);
    row[field] = v;
  }
  const url = editIndex !== null ? `/api/${TABLE}/${editIndex}` : `/api/${TABLE}`;
  const method = editIndex !== null ? "PUT" : "POST";
  await fetch(url, { method, headers: {"Content-Type": "application/json"}, body: JSON.stringify(row) });
  closeModal();
  await loadData();
}

async function confirmDelete(idx) {
  const pk = tableSchema.pk;
  const label = pk ? rows[idx][pk] : `row #${idx}`;
  if (!confirm(`Delete "${label}"?`)) return;
  await fetch(`/api/${TABLE}/${idx}`, { method: "DELETE" });
  await loadData();
}

document.getElementById("search").addEventListener("input", render);

function toggleTheme() {
  const dark = document.getElementById("theme-toggle").checked;
  const theme = dark ? "dark" : "light";
  document.documentElement.setAttribute("data-theme", theme);
  localStorage.setItem("theme", theme);
}

(function() {
  var t = localStorage.getItem("theme") ||
    (window.matchMedia("(prefers-color-scheme: dark)").matches ? "dark" : "light");
  document.documentElement.setAttribute("data-theme", t);
  if (t === "dark") document.getElementById("theme-toggle").checked = true;
})();

loadData();
